### PR TITLE
Speed up IN list types analysis

### PR DESCRIPTION
**Pull request by wendigo**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
https://github.com/trinodb/trino/issues/5950

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -2574,22 +2567,34 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
         {
             // determine super type
             Type superType = UNKNOWN;
+
+            ListMultimap<Type, Expression> typeExpressions = LinkedListMultimap.create();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2571]: `ArrayListMultimap` should have better perf

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -2574,22 +2567,34 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
             // verify all expressions can be coerced to the superType
-            for (Expression expression : expressions) {
-                Type type = process(expression, context);
+            for (Type type : types) {
+                List<Expression> coercionCandidates = typeExpressions.removeAll(type);
+
                 if (!type.equals(superType)) {
                     if (!typeCoercion.canCoerce(type, superType)) {
-                        throw semanticException(TYPE_MISMATCH, expression, message, superType);
+                        throw semanticException(TYPE_MISMATCH, typeExpressions.get(type).get(0), message, superType);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2593]: this `typeExpressions.get(type).get(0)` won't work as you have just removed all elements for type

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -2598,13 +2603,23 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
-            NodeRef<Expression> ref = NodeRef.of(expression);
-            expressionCoercions.put(ref, superType);
-            if (typeCoercion.isTypeOnlyCoercion(type, superType)) {
-                typeOnlyCoercions.add(ref);
+            addOrReplaceExpressionCoercion(expression, superType, typeCoercion.isTypeOnlyCoercion(type, superType));
+        }
+
+        private void addOrReplaceExpressionCoercion(Expression expression, Type superType, boolean isTypeOnlyCoercion)
+        {
+            addOrReplaceExpressionsCoercion(List.of(expression), superType, isTypeOnlyCoercion);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2611]: `List.of(expression)` -> `ImmutableList.of(expression)`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -2598,13 +2603,23 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
+        }
+
+        private void addOrReplaceExpressionsCoercion(List<Expression> expressions, Type superType, boolean isTypeOnlyCoercion)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2614]: you don't need to pass `boolean isTypeOnlyCoercion` as argument since this method is already batching expressions (no perf gain)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -2574,22 +2567,34 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
+            for (Type type : types) {
+                Optional<Type> newSuperType = typeCoercion.getCommonSuperType(superType, type);
                 if (newSuperType.isEmpty()) {
-                    throw semanticException(TYPE_MISMATCH, expression, message, superType);
+                    throw semanticException(TYPE_MISMATCH, typeExpressions.get(type).get(0), message, superType);
                 }
                 superType = newSuperType.get();
             }
 
             // verify all expressions can be coerced to the superType
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2587]: side note: I don't recall why w need to check this after we determine super type. @martint @kasiafi  do you recall why we do it?

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 6]
```diff
@@ -45,6 +53,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<NodeRef<Expression>, Map<NodeRef<Expression>, Type>> expressionTypesCache = buildNonEvictableCache(
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:56]: preserve `ExpressionAnalyzer` instead (it caches types internally). It will also work with nested expressions compared to `expressionTypesCache` here

File: core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java [hunk 7]
```diff
@@ -69,11 +68,11 @@ public PlanNode defaultRewrite(PlanNode node)
          */
         public PlanNode defaultRewrite(PlanNode node, C context)
         {
-            List<PlanNode> children = node.getSources().stream()
-                    .map(child -> rewrite(child, context))
-                    .collect(toImmutableList());
-
-            return replaceChildren(node, children);
+            ImmutableList.Builder<PlanNode> children = ImmutableList.builderWithExpectedSize(node.getSources().size());
+            for (PlanNode child : node.getSources()) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java:72]: maybe
```
node.getSources().forEach(source -> children.add(rewrite(child, context)));
```

File: testing/trino-tests/src/test/java/io/trino/tests/TestLocalQueries.java [hunk 8]
```diff
@@ -117,4 +117,17 @@ public void testTransformValuesInTry()
                         "FROM (VALUES map(ARRAY[1, 2], ARRAY[0, 0]),  map(ARRAY[28], ARRAY[2]), map(ARRAY[18], ARRAY[2]), map(ARRAY[4, 5], ARRAY[1, 0]),  map(ARRAY[12], ARRAY[3])) AS t(m)",
                 "VALUES NULL, '{\"28\":14}', '{\"18\":9}', NULL, '{\"12\":4}'");
     }
+
+    @Override
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/tests/TestLocalQueries.java:121]: test is missing

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 9]
```diff
@@ -2574,22 +2567,34 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
-            for (Expression expression : expressions) {
-                Type type = process(expression, context);
+            for (Type type : types) {
+                List<Expression> coercionCandidates = typeExpressions.removeAll(type);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2589]: this list could be huge so I wanted to free memory as soon as possible

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2589]: > this list could be huge so I wanted to free memory as soon as possible

I don't think it's a problem though, expressions are already there and with `ArrayListMultimap` you would keep just references to objects. Let's not pre-optimize

**Review by kasiafi — COMMENTED** [review]
I have some thoughts on the `ExpressionAnalyzer` part. Please give me a chance to comment, I'll hopefully get back to it tomorrow.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 4 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java]: Done

**Comment by wendigo** [comment]
Together with https://github.com/trinodb/trino/pull/11902 it yields following performance improvement:

Screenshot 2022-04-12 at 18 36 45

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 10]
```diff
@@ -2568,22 +2560,32 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
             Type superType = UNKNOWN;
+
+            ListMultimap<Type, Expression> typeExpressions = ArrayListMultimap.create();
             for (Expression expression : expressions) {
-                Optional<Type> newSuperType = typeCoercion.getCommonSuperType(superType, process(expression, context));
+                typeExpressions.put(process(expression, context), expression);
+            }
+
+            // We need an explicit copy to avoid ConcurrentModificationException
+            Set<Type> types = typeExpressions.keySet();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2570]: i don't see explicit copy here.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 11]
```diff
@@ -2592,13 +2594,22 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
         private void addOrReplaceExpressionCoercion(Expression expression, Type type, Type superType)
         {
-            NodeRef<Expression> ref = NodeRef.of(expression);
-            expressionCoercions.put(ref, superType);
+            addOrReplaceExpressionsCoercion(List.of(expression), type, superType);
+        }
+
+        private void addOrReplaceExpressionsCoercion(List<Expression> expressions, Type type, Type superType)
+        {
+            Map<NodeRef<Expression>, Type> expressionRefs = expressions
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2602]: "expressionRefs"?

(the name could be OK for a list, but that's a Map)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 12]
```diff
@@ -2592,13 +2594,22 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
+                    .stream()
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2603]: per code style, this belongs to prev line

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 13]
```diff
@@ -2592,13 +2594,22 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
 
         private void addOrReplaceExpressionCoercion(Expression expression, Type type, Type superType)
         {
-            NodeRef<Expression> ref = NodeRef.of(expression);
-            expressionCoercions.put(ref, superType);
+            addOrReplaceExpressionsCoercion(List.of(expression), type, superType);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2597]: "batch" version of `addOrReplaceExpressionsCoercion` looks complex and unnecessary

i would rather keep the old method `addOrReplaceExpressionCoercion(Expression expression, Type type, Type superType)` as is, and just call it in the loop in the only place where you would call it with many expressions.

Am i missing anything?

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 14]
```diff
@@ -45,6 +55,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache = buildNonEvictableCache(
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: `ExpressionAnalyzer` is a ton of mutable state, it's totally insuitable for caching & sharing.

**Review by findepi — DISMISSED** [review]
"Cache expensive TypeAnalyzer instance creation and reduce type analys…
…is cost"

File: core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java [hunk 15]
```diff
@@ -69,11 +68,9 @@ public PlanNode defaultRewrite(PlanNode node)
          */
         public PlanNode defaultRewrite(PlanNode node, C context)
         {
-            List<PlanNode> children = node.getSources().stream()
-                    .map(child -> rewrite(child, context))
-                    .collect(toImmutableList());
-
-            return replaceChildren(node, children);
+            ImmutableList.Builder<PlanNode> children = ImmutableList.builderWithExpectedSize(node.getSources().size());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java:71]: Can you move it to separate PR? I'd like to merge this.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 13 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2597]: I wanted to reduce the number of `typeCoercion.isTypeOnlyCoercion(type, superType)` calls, hence the batching.

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java (hunk 14 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: That was proposed by @sopel39 . My previous version was actually caching only `Map<RefNode<Expression>, Type>`

File: core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java (hunk 15 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/planner/plan/SimplePlanRewriter.java:71]: https://github.com/trinodb/trino/pull/11919

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java (hunk 14 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: `Type` is immutable, so it was definitely a better idea 

(not judging whether this particular cache in this place is required)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 13 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2597]: Then maybe overload, allowing the caller to provide this information?

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java (hunk 14 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: @sopel39 wdyt?

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: > ExpressionAnalyzer is a ton of mutable state, it's totally insuitable for caching & sharing.

I don't think it's a problem since it only adds new information to analysis. It already caches via `io.trino.sql.analyzer.ExpressionAnalyzer.Visitor#process`. Since we only care about types, we could maybe prime new `ExpressionAnalyzer` instance with `Map<NodeRef<Expression>, Type>`, but I'm not sure it's worth extra complexity.

> That was proposed by @sopel39 . My previous version was actually caching only Map<RefNode<Expression>, Type>

IIRC that map was not passed to `ExpressionAnalyzer` hence it wouldn't work with sub-expressions

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: Anyway, for me planning is a process so keeping `ExpressionAnalyzer` is not really an issue. Actually, it's probably desirable since I want to cache `ExpressionInterpreter` too and it needs to be able to fetch type information for new expressions.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 16]
```diff
@@ -2598,13 +2603,23 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
+        {
+            Map<NodeRef<Expression>, Type> expressionRefs = expressions.stream().map(expression -> Map.entry(NodeRef.of(expression), superType)).collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+            expressionCoercions.putAll(expressionRefs);
+            if (isTypeOnlyCoercion) {
+                typeOnlyCoercions.addAll(expressionRefs.keySet());
             }
-            else if (typeOnlyCoercions.contains(ref)) {
-                typeOnlyCoercions.remove(ref);
+            else {
+                expressionRefs.keySet().forEach(typeOnlyCoercions::remove);
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2612]: This is needed in the case when we update a coercion for an expression, and we need to update the `typeOnlyCoercion` property accordingly.

From the original commit message:
```
Fix coercion overrides in ExpressionAnalyzer

The comparison which has type-only in-list (eg decimal 2.0, decimal 30.0) and double value throws exception. The cause is the expression remains type-only even though decimal and double are not type-only relation. This change is removing the expression from typeOnlyCoercions when the value no longer become type-only.
```

Do not remove this logic. Even if you flatten processing of IN to a single pass, and this is no more needed for IN, there are still other usages of this method.

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 17]
```diff
@@ -87,4 +101,73 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
...
+    {
+        return session.getTransactionId()
+                .map(TransactionId::toString)
+                .orElse(session.getQueryId().getId());
+    }
+
+    private static class CacheIdentityKey
+    {
+        private final String id;
+        private final Iterable<Expression> expressions;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:135]: The  is mutable, so it's not a good candidate for a cache key.
Did you mean to materialize the list with `ImmutableList.copyOf` ?

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 18]
```diff
@@ -87,4 +101,73 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
...
+    }
+
+    private static String getIdFromSession(Session session)
+    {
+        return session.getTransactionId()
+                .map(TransactionId::toString)
+                .orElse(session.getQueryId().getId());
+    }
+
+    private static class CacheIdentityKey
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:132]: Can you describe the meaning of the class's name?

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 19]
```diff
@@ -87,4 +101,73 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
...
+            }
+
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            CacheIdentityKey that = (CacheIdentityKey) o;
+
+            return Objects.equals(id, that.id)
+                    && Objects.equals(expressions, that.expressions);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:164]: Expressions should not be compared by equality.
Are you sure it is safe to do so?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 20]
```diff
@@ -2592,13 +2594,22 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
...
-            expressionCoercions.put(ref, superType);
+            addOrReplaceExpressionsCoercion(List.of(expression), type, superType);
+        }
+
+        private void addOrReplaceExpressionsCoercion(List<Expression> expressions, Type type, Type superType)
+        {
+            Map<NodeRef<Expression>, Type> expressionRefs = expressions
+                    .stream()
+                    .map(expression -> Map.entry(NodeRef.of(expression), superType))
+                    .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2605]:
```suggestion
                    .collect(toImmutableMap(NodeRef::of, expression -> superType));
```

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 21]
```diff
@@ -2219,14 +2222,12 @@ protected Type visitInPredicate(InPredicate node, StackableAstVisitorContext<Con
             }
 
             Type declaredValueType = process(value, context);
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:2224]: This can now be moved to the `else` branch.

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java (hunk 19 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:164]: I strongly opt for caching `Map<RefNode, Type>`, as that's what you actually need here.

The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner. It is intended for the Analysis phase, where it provides different kind of information to the Planner via the Analysis object. If you need to analyze expressions later for some reason, there's the `analyzeExpressions` method, and `createConstantAnalyzer` method, depending on your use case. Those methods give you access to the "analyzing" capability of the `ExpressionAnalyzer` while they hide the complexity which is not relevant at that point.

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java (hunk 14 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: I strongly opt for caching `Map<RefNode, Type>`, as that's what you actually need here.

The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner. It is intended for the Analysis phase, where it provides different kind of information to the Planner via the Analysis object. If you need to analyze expressions later for some reason, there's the `analyzeExpressions` method, and `createConstantAnalyzer` method, depending on your use case. Those methods give you access to the "analyzing" capability of the `ExpressionAnalyzer` while they hide the complexity which is not relevant at that point.

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:58]: > The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner.

This is exactly what is happening right now. Every `io.trino.sql.planner.TypeAnalyzer#getTypes` call creates a new `ExpressionAnalyzer` instance (via `analyzeExpressions`) and does a full analysis. Overall it's expensive if done repeatedly.
We want to reduce that cost by keeping `ExpressionAnalyzer` instance during planning, which should be fine since planning is a process rather than moving from one immutable state to the other.

> I strongly opt for caching Map<RefNode, Type>, as that's what you actually need here.

I would like to keep `ExpressionInterpreter` too. For that I need a utility to return `type` for a (sub)expression rather than analyzing entire expression to get full `Map<NodeRef<Expression>, Type>` map. In that context keeping `ExpressionAnalyzer` is more natural.

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 22]
```diff
@@ -45,6 +55,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache = buildNonEvictableCache(
+            CacheBuilder.newBuilder()
+                    // Try to evict entries as soon as possible to keep cache relatively small
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:60]: IMO `TypeAnalyzer` should be created per query instance and possibly extracted as an interface. It should be then created and used in `io.trino.sql.planner.LogicalPlanner#plan`. No `NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache` is needed then

File: testing/trino-tests/src/test/java/io/trino/tests/TestLocalQueries.java [hunk 23]
```diff
@@ -117,4 +117,17 @@ public void testTransformValuesInTry()
...
+    protected Object[][] largeInValuesCountData()
+    {
+        return new Object[][] {
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/tests/TestLocalQueries.java:124]: We also need to add a case where `IN` list is in subexpression, e.g: `x IS NULL or y IN (...)`

File: core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java [hunk 24]
```diff
@@ -54,17 +72,13 @@ public TypeAnalyzer(PlannerContext plannerContext, StatementAnalyzerFactory stat
...
-                new AllowAllAccessControl(),
-                inputTypes,
-                expressions,
-                ImmutableMap.of(),
-                WarningCollector.NOOP,
-                OTHERS)
-                .getExpressionTypes();
+        try {
+            ExpressionAnalyzer analyzer = analyzerCache.get(new CacheIdentityKey(getIdFromSession(session), expressions), () -> createExpressionAnalyzer(session, plannerContext, statementAnalyzerFactory, inputTypes));
+            return analyzeExpressions(analyzer, expressions).getExpressionTypes();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java:77]: Just call `ExpressionAnalyzer#analyze` directly, no need to go through `ExpressionAnalysis`

**Comment by kasiafi** [comment]
> This is exactly what is happening right now. Every io.trino.sql.planner.TypeAnalyzer#getTypes call creates a new ExpressionAnalyzer instance (via analyzeExpressions) and does a full analysis. Overall it's expensive if done repeatedly.

Wouldn't caching `Map<NodeRef<Expression>, Type>` solve the issue? We could even entirely avoid re-analyzing expressions if we captured all the expressionTypes established by the ExpresionAnalyzer in the Analysis phase, and initially fed the cache with them.
There could be some nuance around the NodeRef though, if the expression was for some reason swapped for identical copy.

> For that I need a utility to return type for a (sub)expression rather than analyzing entire expression to get full Map<NodeRef<Expression>, Type> map.

Type for a (sub)expression would be there in the cache.

> I would like to keep ExpressionInterpreter too.

Why? Is that to avoid re-evaluating constant expressions? We could cache `Expression -> Object`.

**Comment by sopel39** [comment]
> Type for a (sub)expression would be there in the cache.

Why would it be in cache? It can be a new instance of expression (e.g. symbol changes). Bringing new instance of `ExpressionAnalyzer` just to analyze an expression seems like an overkill.

> Why? Is that to avoid re-evaluating constant expressions? We could cache Expression -> Object.

Yes. We do that repetitively in planning (e.g. with `simplifyExpressions`). Why to externalize cache from `io.trino.sql.planner.ExpressionInterpreter` rather than keeping `io.trino.sql.planner.ExpressionInterpreter` instance? Not that `io.trino.sql.planner.ExpressionInterpreter` already has some cache internally

**Comment by kasiafi** [comment]
In my opinion, keeping the ExpressionAnalyzer is much more acceptable than creating and caching many ExpressionAnalyzers.

However, I still consider it kind of "abstraction leak". The ExpressionAnalyzer is here to perform correctness checks, determine coercions, etc on the pre-planning phase. Computing types is kind of "implementation detail". We learned to reuse this capability throughout the Planner, because we found that we need to know the types over and over. The new IR should definitely be enhanced with types (and also with pre-computed constants).

While we currently reuse the ExpressionAnalyzer for getting the expression types, we should try to isolate this capability from all the other work that ExpressionAnalyzer does. For that reason, we use the public methods rather than the constructor, which also involves irrelevant parts, like the Analysis.

**Comment by kasiafi** [comment]
I'm OK with having a persistent (per query) ExpressionInterpreter with caching.
Regarding the ExpressionAnalyzer, we could extract just the part we need, that is - pure type analysis and the cache, and let it live through the Planning phase.

**Comment by kasiafi** [comment]
>@kasiafi What does it mean in practice? You mean you would like to have something like ExpressionTypeAnalyzer that just populates io.trino.sql.analyzer.ExpressionAnalyzer#expressionTypes? Is it practical to split ExpressionAnalyzer like that?

Yes, that was my thinking.
It might not be practical, because in the `ExpressionAnalyzer` the type analysis logic is intermixed with correctness checks, recording coercions, subqueries etc. Also, some kinds of expressions, e.g. SubqueryExpression, should not be present at later phases of planning.

However, I still think that caching Expression -> Type should be sufficient, and occasional creation of new ExpressionAnalyzer is better than "pulling" it throughout the Optimizer. And to make it the most "occasional", we should consider preserving the NodeRefs whenever possible (mostly, in ExpressionInterpreter) instead of creating identical copies of Expressions.

**Comment by sopel39** [comment]
> However, I still think that caching Expression -> Type should be sufficient, and occasional creation of new ExpressionAnalyzer is better than "pulling" it throughout the Optimizer. 

Why it would be better? You end up performing all these correctness checks anyway, so why to pretend we don't use `ExpressionAnalyzer`? Is splitting `ExpressionAnalyzer` something we could do later?

> And to make it the most "occasional", we should consider preserving the NodeRefs whenever possible (mostly, in ExpressionInterpreter) instead of creating identical copies of Expressions.

I don't like that approach mostly because `getType` call becomes potentially expensive.

**Comment by findepi** [comment]
I dismissed my review. @martint ptal if you happen to have time.

**Comment by bitsondatadev** [comment]
:wave: @wendigo - this PR has become inactive. If you're still interested in working on it, please let us know.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
